# Overview
# Author: Erika CastaÃ±eda  
# Date: 10 Dec 2023
#
# This R script analyzes the statistics of a classification task.
# It processes classification results stored in a CSV file, checks for chromosome classifications,
# and generates visualizations using ggplot2. The script integrates data from MUMmer and chromosome 
# information to visualize classification statistics.

## Libraries Used
library(ggplot2)  # For data visualization
library(dplyr)    # For data manipulation and transformation
library(readr)    # For reading CSV and text files

## Data Input 

# The data for this analysis originates from the output generated by MUMmer 4.0.0beta2.
# MUMmer is used to map a de novo assembled genome against the reference genome, with the goal of
# classifying contigs at the chromosome level.

# **Reference Genome Information:**
# 
# - **Genome Name**: Gallus gallus bGalGal1.mat.broiler.GRCg7b
# - **Reference**: This reference genome is specific to the WZ system for Gallus gallus (chicken).
# - **Source**: This genome assembly provides detailed information about chicken chromosomes and contigs.

# **Data File Descriptions:**
# 
# 1. **Classification Results (CSV File)**:
#    - File Path: /home/pc-cito/chicken/resum_classification.csv
#    - Columns:
#      - X1 > Chromosome: character
#      - X2 > Contig: character
#      - X3 > Start position: numeric
#      - X4 > End position: numeric

# Read classification results from CSV file
## Data Input and Preprocessing
df <- read_table("/resum_classification.csv", col_names = FALSE) %>%
  rename(Chromosome = X1, Contig = X2, Start = X3, End = X4)

# Clean up 'Chromosome' column
df$`RefSeq seq accession` <- sub(">", "", df$Chromosome)

## Reference Database
# Information about the reference genome:
# 
# - File Path: /home/pc-cito/chicken/sequence_report.tsv
# - Format: TSV (Tab-Separated Values)
# - Columns:
#   - Chromosome name: character
#   - RefSeq seq accession: character

# Read chromosome information
chromAlias <- read_delim("/sequence_report.tsv", 
                         delim = "\t", escape_double = FALSE, 
                         trim_ws = TRUE)

## Data Analysis
# Merge classification results with chromosome information
result <- merge(chromAlias, df, by = "RefSeq seq accession", all.y = TRUE)
result <- result[, c('Chromosome name', 'RefSeq seq accession', 'Contig', 'Start', 'End')]

# Check for unclassified chromosomes
vacios <- filter(result, is.na(`Chromosome name`))

## Data Visualization
# Summarize data for visualization
resumen <- result %>%
  group_by(`Chromosome name`, Contig) %>%
  summarise(n = n()) %>%
  ungroup()

# Identify exclusive contigs (those appearing in only one chromosome)
contigs_exclusivos <- resumen %>%
  group_by(Contig) %>%
  filter(n_distinct(`Chromosome name`) == 1 & n() == 1) %>%
  ungroup() %>%
  mutate(es_exclusivo = TRUE)

resumen <- resumen %>%
  left_join(contigs_exclusivos %>%
              select(Contig, es_exclusivo), by = "Contig") %>%
  mutate(es_exclusivo = ifelse(is.na(es_exclusivo), FALSE, TRUE))

# Create barplot to visualize unique vs duplicate contigs
etiquetas_numericas <- resumen %>%
  group_by(`Chromosome name`, es_exclusivo) %>%
  summarise(num_contigs = sum(n)) 

ggplot(etiquetas_numericas, aes(x = `Chromosome name`, y = num_contigs, fill = factor(es_exclusivo))) +
  geom_bar(stat = "identity", position = "dodge") +
  geom_text(aes(label = num_contigs), position = position_dodge(width = 0.9), vjust = -0.5) +
  scale_fill_manual(values = c('brown3', "green4"), name = NULL) +
  labs(title = "Barplot of Unique vs Duplicate Contigs",
       x = "Chromosomes",
       y = "Number of Contigs") +
  theme_bw()

# Save cleaned dataset of exclusive contigs
classified <- filter(resumen, es_exclusivo == TRUE)
write_csv(classified, "clean_classified.csv")
